logic/Rule.java

```diff
@@ -169,7 +169,7 @@ public final int hashCode() {
     private Set<HeadConcludable<?, ?>> createHead(Set<Constraint> thenConstraints, Set<Variable> constraintContext) {
         HashSet<HeadConcludable<?, ?>> thenConcludables = new HashSet<>();
         thenConstraints.stream().filter(Constraint::isThing).map(Constraint::asThing)
-                .flatMap(constraint -> HeadConcludable.of(constraint, constraintContext).stream()).forEach(thenConcludables::add);
+                .map(constraint -> HeadConcludable.of(constraint, constraintContext)).forEach(thenConcludables::add);
```

Comment:
We've gotten rid of generalisations, so the output of `.of` is a single HeadConcludable rather than a Set or Stream.